### PR TITLE
7.0 change public holiday

### PR DESCRIPTION
**Pull request by OSguard**
Add country and states
Resubmit of https://code.launchpad.net/~initos.com/openerp-hr/7.0-change-public-holiday/+merge/224610

### CODE REVIEW

File: hr_public_holidays/hr_public_holidays.py [hunk 1]
```diff
@@ -32,39 +34,99 @@ class hr_holidays(osv.osv):
...
+                                         ('country_id', '=', False)]
+
+        ph_ids = self.search(cr, uid, holidays_filter,
             context=context)
+
         if len(ph_ids) == 0:
             return False
 
-        for line in ph_obj.browse(cr, uid, ph_ids[0], context=context).line_ids:
+        if employee_id == None:
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays.py]: You can rework this in this way:

```
states_filter = [('holidays_id', 'in', ph_ids)]
if not employee or not employee.address_id.state_id:
    states_filter.append(('state_ids', '=', False))
else:
    states_filter += ['|', ('state_ids', '=', False), ('state_ids.id', '=', employee.address_id.state_id.id))
```

File: hr_public_holidays/hr_public_holidays.py [hunk 2]
```diff
@@ -32,39 +34,99 @@ class hr_holidays(osv.osv):
...
+                                               ph_ids),
+                                              ('state_ids', '=', False)]
+            else:
+                states_filter = [('holidays_id', 'in',
+                                                ph_ids),
+                                              '|', ('state_ids.id', '=',
+                                              employee.address_id.state_id.id),
+                                              ('state_ids', '=', False)]
+
+        hr_holiday_public_line_obj = self.pool.get('hr.holidays.public.line')
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays.py]: Please use `self.pool['hr.holidays.public.line']`.

File: hr_public_holidays/hr_public_holidays.py [hunk 3]
```diff
@@ -32,39 +34,99 @@ class hr_holidays(osv.osv):
...
                 return True
 
         return False
 
-    def get_holidays_list(self, cr, uid, year, context=None):
+    def get_holidays_list(self, cr, uid, year, employee_id=None, context=None):
 
         res = []
-        ph_ids = self.search(cr, uid, [('year', '=', year)], context=context)
+        if employee_id == None:
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays.py]: Redo in the same way as above.

File: hr_public_holidays/hr_public_holidays_view.xml [hunk 4]
```diff
@@ -32,14 +34,28 @@
...
                     <field name="variable"/>
                 </tree>
             </field>
         </record>
         
+        <record id="view_public_holiday_line_form" model="ir.ui.view">
+            <field name="name">hr.holidays.public.line.form</field>
+            <field name="model">hr.holidays.public.line</field>
+            <field name="arch" type="xml">
+					<form string="Public Holidays">
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays_view.xml]: Indent this correctly (with spaces)

**Comment by bwrsandman** [comment]
Hi @OSguard, I restarted the [travis build](https://travis-ci.org/OCA/hr/builds/30533645) after merging #10, there are quite a few pep8 errors introduced with this PR and they will need to be fixed before we can merge your code.

```
LINE_LENGTH=120

flake8 . --max-line-length=${LINE_LENGTH} --filename=__init__.py --ignore=F401 --exclude=__unported__

status1=$?

flake8 . --max-line-length=${LINE_LENGTH} --exclude=__unported__,__init__.py

./hr_public_holidays/hr_public_holidays.py:48:24: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:52:47: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:56:43: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:57:42: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:60:13: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:65:24: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:67:45: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:68:45: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:70:45: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:72:48: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:73:47: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:76:49: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:77:47: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:78:47: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:83:41: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:93:24: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:97:47: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:101:43: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:102:42: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:108:24: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:110:44: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:111:43: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:113:45: E711 comparison to None should be 'if cond is None:'
./hr_public_holidays/hr_public_holidays.py:115:48: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:116:47: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:119:49: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:120:47: E127 continuation line over-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:121:47: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:126:41: E128 continuation line under-indented for visual indent
./hr_public_holidays/hr_public_holidays.py:143:121: E501 line too long (130 > 120 characters)
```

**Comment by OSguard** [comment]
now it should be fine

**Comment by dreispt** [comment]
Just a question: 
I see holidays can be defined per country/state.
Are city holidays unusual in most countries?

**Comment by OSguard** [comment]
i can only say in germany it depends on the state.

File: hr_public_holidays/hr_public_holidays.py [hunk 5]
```diff
@@ -32,39 +34,88 @@ class hr_holidays(osv.osv):
...
+
         if len(ph_ids) == 0:
             return False
 
-        for line in ph_obj.browse(cr, uid, ph_ids[0], context=context).line_ids:
+        states_filter = [('holidays_id', 'in', ph_ids)]
+        if not employee or not employee.address_id.state_id:
+            states_filter.append(('state_ids', '=', False))
+        else:
+            states_filter += ['|', ('state_ids', '=', False), ('state_ids.id', '=', employee.address_id.state_id.id)]
```
Comment (dreispt) [inline hr_public_holidays/hr_public_holidays.py]: Line too long? (PEP8)

File: hr_public_holidays/hr_public_holidays.py [hunk 6]
```diff
@@ -78,6 +129,7 @@ class hr_holidays_line(osv.osv):
         'date': fields.date('Date', required=True),
         'holidays_id': fields.many2one('hr.holidays.public', 'Holiday Calendar Year'),
         'variable': fields.boolean('Date may change'),
+        'state_ids': fields.many2many('res.country.state', 'hr_holiday_public_state_rel', 'line_id', 'state_id', 'Related states')
```
Comment (dreispt) [inline hr_public_holidays/hr_public_holidays.py]: Line too long? (PEP8)

**Comment by dreispt** [comment]
You are not required full PEP8 on an existing module (nor it should be done at the same time with other changes), but new/changed lines should be made PEP8 compliant.

**Comment by bwrsandman** [comment]
@dreispt 7.0 branch is now fully pep8 compliant: [Build Status](https://travis-ci.org/OCA/hr)
Any pep8 errors from this build comes from new code (changed lines)

```
flake8 . --max-line-length=120 --filename=__init__.py --ignore=F401 --exclude=__unported__
flake8 . --max-line-length=120 --exclude=__unported__,__init__.py
```

File: hr_public_holidays/hr_public_holidays_view.xml [hunk 7]
```diff
@@ -32,14 +34,28 @@
             <field name="name">hr.holidays.public.line.tree</field>
             <field name="model">hr.holidays.public.line</field>
             <field name="arch" type="xml">
-                <tree string="Public Holidays" editable="top">
+                <tree string="Public Holidays">
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays_view.xml]: Why do you remove editability? many2many fields has a widget for trees.

File: hr_public_holidays/hr_public_holidays.py [hunk 8]
```diff
@@ -31,40 +33,104 @@ class hr_holidays(osv.osv):
...
+                                  employee.address_id.state_id.id),
+                                 ('state_ids', '=', False)]
+
+        hr_holiday_public_line_obj = self.pool['hr.holidays.public.line']
+        holidays_line_ids = \
+            hr_holiday_public_line_obj.search(cr, uid, states_filter)
+
         [res.append(l.date)
-         for l in self.browse(cr, uid, ph_ids[0], context=context).line_ids]
+         for l in hr_holiday_public_line_obj.browse(cr, uid,
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays.py]: Indentation is not correct

File: hr_public_holidays/hr_public_holidays_view.xml [hunk 9]
```diff
@@ -32,14 +34,28 @@
...
                     <field name="variable"/>
                 </tree>
             </field>
         </record>
         
+        <record id="view_public_holiday_line_form" model="ir.ui.view">
+        	<field name="name">hr.holidays.public.line.form</field>
+            <field name="model">hr.holidays.public.line</field>
+            <field name="arch" type="xml">
+				<form string="Public Holidays">
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays_view.xml]: Indentation incorrect

File: hr_public_holidays/hr_public_holidays.py [hunk 10]
```diff
@@ -31,40 +33,104 @@ class hr_holidays(osv.osv):
...
+
+        ph_ids = self.search(cr, uid, holidays_filter,
+                             context=context)
 
-        ph_obj = self.pool.get('hr.holidays.public')
-        ph_ids = ph_obj.search(cr, uid, [
-            ('year', '=', dt.year),
-        ],
-            context=context)
         if len(ph_ids) == 0:
```
Comment (pedrobaeza) [inline hr_public_holidays/hr_public_holidays.py]: You can put: `if not ph_ids:`

Regards.

**Comment by bwrsandman** [comment]
:+1: 

Unittests are failing but it's due to the aeroo modules #13

**Comment by dreispt** [comment]
:thumbsup:

**Comment by pedrobaeza** [comment]
Well, the module has some round edges, like the state edition on the tree, but for now can work.

Thanks. Regards.

:+1:

**Pull request merged**